relayer/src/client.rs

```diff
@@ -5,116 +5,254 @@ use std::{
-                            receipts.push(receipt)
-                        } else {
-                            log::warn!(
-                                "Could not get a transaction receipt for tx {}",
-                                hash.encode_hex()
-                            );
-                            continue 'outer;
-                        }
-                    }
-
-                    if !receipts.is_empty() {
-                        let json = serde_json::to_string(&receipts)?;
-                        db.insert_receipts(block_hash, &json)?;
-                    } else {
-                        log::debug!(
-                            "Block {} does not have any receipts",
-                            block_hash.encode_hex()
-                        );
-                    }
-                } else {
-                    log::info!(
-                        "Could not get a block by block_hash {}",
-                        block_hash.encode_hex()
-                    );
-                    continue 'outer;
-                }
+    async fn finalization_loop(&mut self) -> Result<()> {
+        let mut latest_finalized_block =
+            self.db.select_latest_block_height(BlockType::Finalized)?;
+        let mut duration = tokio::time::interval(Duration::from_secs(5));
+        loop {
+            exit_if_term(self.term.clone());
+            duration.tick().await;
+            let finalized_block = self
+                .client
+                .get_block_by_number(BlockTag::Finalized, false)
+                .await;
+            let finalized_block = if let Ok(Some(finalized_block)) = finalized_block {
+                finalized_block
+            } else {
+                log::warn!(target: "relayer::client::finalization_loop","Failed to get finalized block, retrying in {} seconds", duration.period().as_secs());
+                tokio::time::sleep(Duration::from_secs(5)).await;
+                continue;
+            };
+
+            let latest_processed_block =
+                self.db.select_latest_block_height(BlockType::Processed)?;
+            if Some(finalized_block.number) == latest_finalized_block
+                && Some(finalized_block.number) == latest_processed_block
+            {
+                log::info!(target: "relayer::client::finalization_loop","No new finalized blocks, retrying in {} seconds", duration.period().as_secs());
+                continue;
+            }
+            log::info!(target: "relayer::client::finalization_loop","New finalized block: {}", finalized_block.number);
+
+            if let Err(e) = self.db.insert_or_update_latest_block_info(
+                BlockType::Finalized,
+                finalized_block.number,
+                H256(finalized_block.hash.0),
+            ) {
+                log::error!(target: "relayer::client::finalization_loop","Failed to update latest finalized block info: {}", e);
+                continue;
```

Comment:
Why `continue`?
Can this operation be successful next time?

Reply:
I'm not an expert on sqllite. Probably not, because it means it corrupted, but if it would be later something bigger (like Postgres), it could be successful.